src/common/src/array/data_chunk.rs

```diff
@@ -95,23 +96,20 @@ impl DataChunk {
     }
 
     /// Build a `DataChunk` with rows.
+    ///
+    /// Panics if the `rows` is empty.
+    ///
+    /// Should prefer using [`DataChunkBuilder`] instead to avoid unnecessary allocation
+    /// of rows.
     pub fn from_rows(rows: &[impl Row], data_types: &[DataType]) -> Self {
-        let mut array_builders = data_types
-            .iter()
-            .map(|data_type| data_type.create_array_builder(1))
-            .collect::<Vec<_>>();
+        let mut builder = DataChunkBuilder::new(data_types.to_vec(), rows.len() + 1);
 
         for row in rows {
-            for (datum, builder) in row.iter().zip_eq_debug(array_builders.iter_mut()) {
-                builder.append(datum);
-            }
+            let none = builder.append_one_row(row);
+            debug_assert!(none.is_none());
         }
 
-        let new_columns = array_builders
-            .into_iter()
-            .map(|builder| builder.finish().into())
-            .collect::<Vec<_>>();
-        DataChunk::new(new_columns, rows.len())
+        builder.consume_all().expect("empty chunk")
```

Comment:
It'd be better to write expect message like `builder should contain at least one row`.